smtp/CMakeLists.txt

```diff
@@ -4,7 +4,90 @@ endif(APPLE)
+# smtpc_init()) for loading system's libcurl.so and everything
+# should be fine? Sadly, no...
+#
+# System's libcurl.so is usually built without
+# -Wl,-Bsymbolic-functions (the exception is Ubuntu -- it
+# surprisingly differs here from Debian) and dynamic relocations
+# are generated for exported libcurl functions that libcurl calls
+# from its code. Those calls are dynamic and an application can
+# retarget them just by exporting the target function from its
+# executable. A user may retarget those calls using LD_PRELOAD.
+# This mechanism aims to provide a way to extend or change a
+# shared library functionality.
+#
+# However at least some of such calls (from libcurl's code to a
+# libcurl's exported function) are internal by its nature and,
+# AFAIU, must land to the same libcurl.so version / build. The
+# example I found is calling of curl_multi_add_handle() from the
+# curl_easy_perform() implementation. It fails for me when
+# system's libcurl.so (7.75.0 with smtp(s) support) processes an
+# smtp request and calls curl_multi_add_handle() from tarantool
+# 2.5.3 (libcurl 7.71.1 without smtp(s) support). The error is
+# CURLE_FAILED_INIT.
+#
+# Back to the main question. What is going on if we'll add curl
+# library to the NEEDED entry of the module's dynamic library?
+#
+# Tarantool is loaded and calls require('smtp'), the module's
+# lib.so library starts to load. The dynamic linker find the
+# NEEDED entry with system's libcurl.so and loads it. The
+# libcurl.so library has dynamic relocations and they are resolved
+# to the executable's symbols. The module loads libcurl.so using
+# dlopen(). The resolved dynamic relocations remain resolved
+# to the executable and the RTLD_DEEPBIND flag does not affect
+# them. Due to differences in versions / supported protocols (I
+# don't know a presice reason) curl_easy_perform() from the loaded
+# libcurl.so becomes broken and always returns CURLE_FAILED_INIT.
+#
+# What if we'll not add the curl library here?
+#
+# If we're lucky and system's libcurl.so is not loaded yet, our
+# dlopen() with RTLD_DEEPBIND will work as expected and all
+# symbols from the library will be resolved to the library itself.
+#
+# Sure, if another module has system's libcurl.so in the NEEDED
+# entry or loads it without RTLD_DEEPBIND and this another module
+# is loaded prior to this one we again in the trouble.
+#
+# There are several facts that more or less mitigates the problem:
+#
+# 1. I don't know any other tarantool's module that uses libcurl
+#    (except built-in http.client, which uses built-in libcurl,
+#    and so irrelevant here).
+# 2. Tarantool offers suitable libcurl implementation since
+#    1.10.10, 2.6.3, 2.7.2, 2.8.1 and in all 2.9+. The module
+#    will use the built-in libcurl when possible (it is not
+#    implemented yet, tracked by gh-24).
+#
+# [1]: https://github.com/tarantool/smtp/issues/29
+# [2]: https://github.com/tarantool/smtp/issues/24
```

Comment:
Wow! Good research!

Reply:
Thanks! I plan to update several details soon.

Reply:
I clarified it a bit with mention of the reference counting that dlopen() performs. This way it should be easier to understand how all those things work.